metaflow/subprocess_manager.py

```diff
@@ -0,0 +1,330 @@
+            # returns when process has been started,
+            # not when it is finished...
+            self.process = await asyncio.create_subprocess_exec(
+                *self.command,
+                cwd=self.cwd,
+                env=self.env,
+                stdout=open(stdout_logfile, "w"),
+                stderr=open(stderr_logfile, "w"),
+            )
+
+            self.log_files["stdout"] = stdout_logfile
+            self.log_files["stderr"] = stderr_logfile
+
+            self.run_called = True
+            return self.process
+        except Exception as e:
+            print("Error starting subprocess: %s" % e)
+            self.cleanup()
+
+    async def stream_logs(
+        self,
+        stream: str,
+        position: Optional[int] = None,
+        timeout_per_line: Optional[float] = None,
+        log_write_delay: float = 0.01,
+    ):
+        """Stream logs from the subprocess using the log files"""
+
+        if self.run_called is False:
+            raise ValueError("No command run yet to get the logs for...")
+
+        if stream not in self.log_files:
+            raise ValueError(
+                "No log file found for '%s', valid values are: %s"
+                % (stream, list(self.log_files.keys()))
+            )
+
+        log_file = self.log_files[stream]
+
+        with open(log_file, mode="r") as f:
+            if position is not None:
+                f.seek(position)
+
+            while True:
+                # wait for a small time for complete lines to be written to the file
+                # else, there's a possibility that a line may not be completely written when
+                # attempting to read it. This is not a problem, but improves readability.
+                await asyncio.sleep(log_write_delay)
+
+                try:
+                    if timeout_per_line is None:
+                        line = f.readline()
+                    else:
+                        line = await asyncio.wait_for(f.readline(), timeout_per_line)
+                except asyncio.TimeoutError as e:
+                    raise LogReadTimeoutError(
+                        "Timeout while reading a line from the log file for the stream: %s"
+                        % stream
+                    ) from e
+
+                # when we encounter an empty line
+                if not line:
+                    # either the process has terminated, in which case we want to break
+                    # and stop the reading process of the log file since no more logs
+                    # will be written to it
+                    if self.process.returncode is not None:
+                        break
+                    # or the process is still running and more logs could be written to
+                    # the file, in which case we continue reading the log file
+                    else:
+                        continue
+
+                position = f.tell()
+                yield position, line.strip()
```

Comment:
I would not do a strip. Maybe rstrip but indentation may be relevant to the user.

Reply:
done